Go 1.23 support

aykevl opened the pull request:
Still a work in progress.

aykevl committed:
ci: use Go 1.23

aykevl committed:
sync/atomic: add And* and Or* compiler intrinsics needed for Go 1.23

aykevl committed:
internal/bytealg: add CompareString

This function was added to Go many years ago, but is starting to be used
in packages in Go 1.23.

aykevl committed:
internal/abi: implement initial version of this package

This package can never be a full version as seen in upstream Go, because
TinyGo is very different. But it is necessary to define so that no code
can accidentally use this package (now or in the future).

It currently defines:

  - NoEscape which is needed by strings.Builder since Go 1.23.
  - FuncPCABI* which is needed by internal/syscall/unix on MacOS.

aykevl committed:
unique: implement custom version of unique package

This version probably isn't as fast as the upstream version, but it is
good enough for now. It also doesn't free unreferenced handles like the
upstream version.

aykevl committed:
runtime: implement timers for Go 1.23

There were a number of changes in time.Timer/time.Ticker that need a
separate implementation for Go 1.22 and Go 1.23.

aykevl committed:
reflect: implement Type.Overflow* functions

They're already implemented for the Value equivalents. But since Go
1.23, such methods also exist for reflect.Type.

aykevl committed:
sync: add Map.Clear()

This was added in Go 1.23 and is needed for the net/mail package.

deadprogram commented:
We're going to need this in the next few weeks when Go 1.23 is released.

deadprogram commented:
@aykevl please note merge conflict needing to be resolved. Also, I think `1.23` release is about to happen.

deadprogram commented:
Go 1.23 has just been released: https://go.dev/dl/#go1.23.0

deadprogram commented:
@aykevl please see https://github.com/tinygo-org/tinygo/pull/4405 specifically the last commit and the resulting CI build failures.

aykevl commented:
Things missing:
 - [x] investigate and fix CI errors
 - [x] range-over-func support
 - [ ] ~`//go:linkname` restrictions~ going to postpone this one

aykevl committed:
syscall: add all MacOS errno values

ELOOP is used starting with Go 1.23. But I figured I could just add the
whole set.

aykevl committed:
compiler: add support for Go 1.23 range-over-func

aykevl commented:
Got a bit further. Let's see how many CI errors remain.

I've implemented range-over-func. I'm going to postpone `//go:linkname` restrictions until a later time, because it's kinda invasive and not critical for Go 1.23 support.

deadprogram commented:
There are several failures in the CI that need to be addressed:

https://github.com/tinygo-org/tinygo/actions/runs/10408226717/job/28825204718?pr=4305
https://github.com/tinygo-org/tinygo/actions/runs/10408226707/job/28825216474?pr=4305
https://github.com/tinygo-org/tinygo/actions/runs/10408226707/job/28825437963?pr=4305

@dgryski @dkegel-fastly @ydnar can you please help @aykevl take a look?

dgryski commented:
I'll take a look today.

aykevl committed:
darwin: work around a linker error for the mime/quotedprintable tests

This is needed for Go 1.23 support.

These functions should ideally get implemented. Until that's done, just
panic here. Apparently panicking in internal/abi.FuncPCABI0 is enough to
fix all linker errors for mime/quotedprintable.

ydnar commented:
🍿

deadprogram commented:
Great work on this @aykevl @dgryski @ydnar thank you very much! Now merging.

QuLogic commented on testdata/errors/types.go:
```diff
@@ -7,6 +7,6 @@ func main() {
 
 // ERROR: # command-line-arguments
-// ERROR: types.go:4:6: a declared and not used
+// ERROR: types.go:4:6: declared and not used: a
```
Note this change does mean tests don't pass on Go < 1.23, though 1.19 is the minimum.

aykevl replied:
Correct. A number of tests only work on the latest Go and LLVM versions. But I think all the tests in main_test.go for example should pass in most or all Go and LLVM versions.
And of course the smoke test pass in all supported Go and LLVM versions, because that's we're testing in CI.

You do have a point though. I've added some basic tests for the oldest supported Go/LLVM version here: https://github.com/tinygo-org/tinygo/pull/4562
